### ISSUE TITLE
Properly allow `string.concat`

### ISSUE THREAD

**Issue by hrkrshnn** [ethereum/solidity#12087]
This may have been allowed unintentionally:

https://github.com/ethereum/solidity/blob/9d6eaa7a46425301ddb961a369b04f211d53691c/libsolidity/ast/Types.cpp#L3817-L3821

https://github.com/ethereum/solidity/blob/9d6eaa7a46425301ddb961a369b04f211d53691c/libsolidity/ast/Types.h#L841

https://github.com/ethereum/solidity/blob/9d6eaa7a46425301ddb961a369b04f211d53691c/libsolidity/ast/Types.h#L866

--- 

Anyway, `string.concat(...)` behaves exactly like `bytes.concat`, i.e., it expects `byte` argument and not string argument. It also returns `bytes memory` and not `string memory`.

Suggestions:

- [ ] Make it return `string memory`
- [ ] Make it accept `string` instead of `bytes`.

**Comment by chriseth** [comment]
We actually intentionally disallow any indexing operations on strings because intuitively, you should be able to access characters of the string by index, but to do that properly, we would have to analyze its utf8 encoding. The same could be said about string concatenation, so maybe it's better to remove it than to change its signature.

**Comment by hrkrshnn** [comment]
Decision: we agreed to allow `string.concat`. The arguments can be `string memory`, `string storage` and string literals.

**Comment by chriseth** [comment]
Is this breaking?

**Comment by axic** [comment]
> Is this breaking?

The return argument change is in my opinion. Making it to accept `string...` input next to `bytes...` is not breaking, but disallowing `bytes...` is breaking again.

**Comment by chriseth** [comment]
Sure, but we could also just say it is a bugfix.

**Comment by cameel** [comment]
I think it's technically breaking but there's zero evidence of actual use in practice so I think treating it as a bug should be ok. It won't really impact anyone in practice.

I did a code search and I can see [zero results for `string.concat` on sourcegraph](https://sourcegraph.com/search?q=context:global+lang:Solidity+content:"string.concat"&patternType=literal&case=yes).
I also tried a few variants with more relaxed criteria (regex, case inensitive, etc.) to make sure it's not missing any cases and all I could find were some instances of `String.concat`, which must be just a library function attached to a struct. I also searched on github, which gives more messy results but after going through first ~20 pages I did not see any actual use of `string.concat`.

**Comment by ekpyron** [comment]
I'd also tend towards making it a bugfix, so we told @nishant-sachdeva to start relative to ``develop`` for now. But we should be able to trivially rebase any code changes relative to ``develop`` to ``breaking``, so we can still change our mind.

**Comment by antazoey** [comment]
I am wondering if I am facing the same (or closely related) bug or if I should open a new issue.

If I have a contract like:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;

contract ShouldFail {
    string public concatres = "";

    function methodB1(bytes32 lolol) public {
        string memory converted_lols = string(abi.encodePacked(lolol));
        string memory result = string.concat(converted_lols, " Captain Janeway");
        concatres = result;
    }
}
```

I can compile it using the `solc` command line compiler tool and I get very strange results.

However, when I paste this same contract into `remix` IDE, it does not compile.

**Comment by hrkrshnn** [comment]
@unparalleled-js This was only released in 0.8.12. It's likely that remix is using a version that's lower than 0.8.12.
